frontend/javascripts/dashboard/dataset/dataset_import_view.js

```diff
@@ -97,28 +102,49 @@ class DatasetImportView extends React.PureComponent<Props, State> {
   async fetchData(): Promise<void> {
     try {
       this.setState({ isLoading: true });
-      const dataset = await getDataset(this.props.datasetId);
+      let dataset = await getDataset(this.props.datasetId);
       let dataSource;
-      let previousDataSource;
       let dataSourceMessages = [];
       if (dataset.isForeign) {
         dataSource = await readDatasetDatasource(dataset);
+        this.setState({ savedDataSourceOnServer: dataSource });
       } else {
+        let savedDataSourceOnServer;
+        let inferredDataSource;
         ({
-          dataSource,
+          dataSource: inferredDataSource,
           messages: dataSourceMessages,
-          previousDataSource,
+          previousDataSource: savedDataSourceOnServer,
         } = await getDatasetDatasource(dataset));
-
-        this.setState({ previousDataSource });
-        if (previousDataSource != null && dataSource != null) {
-          const diff = diffObjects(dataSource, previousDataSource);
-          this.setState({ differenceBetweenDatasources: diff });
+        const didParsingTheSavedDataSourceJSONSucceed =
+          savedDataSourceOnServer != null && savedDataSourceOnServer.status == null;
+        if (didParsingTheSavedDataSourceJSONSucceed) {
+          dataSource = savedDataSourceOnServer;
+        } else {
+          // If the current datasource json is invalid, the inferred version should be used automatically.
+          dataSource = inferredDataSource;
+          this.setState({ didApplyDifferenceBetweenDatasources: true });
+        }
+        this.setState({ savedDataSourceOnServer, inferredDataSource });
```

Comment:
To get rid of those let variables, could you save the `dataSourceMessages` to the state here instead of in line 180? Then you should be able to get rid of the let variable declarations in lines 107, 112, 113 and use `const = ({` in line 114.
Please ignore if this doesn't make sense.